### PR TITLE
Fewer requests on playlist change, bigger buffer

### PR DESCRIPTION
**Pull request by kevinzang**
@cansin @burakbostancioglu @allisonUdemy please review my first attempt to decrease number of requests made by HLS plugin.

https://trello.com/c/P0LBdYB8/4481-2-hls-player-same-chunk-download-bug-3
https://trello.com/c/UKOPoHKn/4482-2-hls-player-random-segment-loads-2
- Previously, whenever the playlist changed (due to variance in user's bandwidth), the plugin reloaded the same segment at the new playlist's resolution. I made it load the next segment. This resolves the multiple requests for the same segment at different resolutions.
- Previously, the plugin limited buffer size to 1 minute's worth of content. Consequently, even for short videos (~3 minutes long), if you play through it once and replay from the beginning, it'll request the segments all over again. I made it only trim the buffer when it detects that the buffer quota is reached. This increases the amount of buffered content that can be stored.
- Previously, if a request for a segment timed out, we would sometimes reload the segment before the one that timed out. I made it only reload the segment that timed out. This was as close as I could get to reproducing the "random segment loads", e.g. `1, 2, 3, 2, 3`, where `3` times out and we needlessly reload `2`. I was never able to get e.g. `1, 3, 2` out of order behavior.

### CODE REVIEW

**Comment by burakbostancioglu** [comment]
LGTM

File: dist/videojs.hls.js [hunk 1]
```diff
@@ -4880,7 +4937,12 @@ videojs.HlsHandler.prototype.fillBuffer = function(mediaIndex) {
   if (mediaIndex === undefined) {
     if (currentBuffered && currentBuffered.length) {
       currentBufferedEnd = currentBuffered.end(0);
-      mediaIndex = this.playlists.getMediaIndexForTime_(currentBufferedEnd);
+      // currentBufferedEnd is rounded up to handle the following case:
+      // suppose we've loaded the first segment, which has a duration of 10s
+      // if currentBufferedEnd is 10s, or slightly more, we correctly realize that we should load
+      // the second segment next
+      // but if it is slightly less than 10s, we try to reload the first segment
```
Comment (allisonUdemy) [inline dist/videojs.hls.js:4944]: Horray for no longer doing this!

**Comment by allisonUdemy** [comment]
Nice commenting in the code!

LGTM

**Comment by cansin** [comment]
I guess LGTM, I don't have any expertise re: HLS honestly.

**Pull request merged**